**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №13

Набор данных: movies.csv

Атрибуты:
+ Название фильма
+ Год выхода
+ Оценка
+ Количество голосов 
+ Длительность 
+ Главный жанр
+ Страна производства 


### 1.Чтение файла (набора данных)

1.1 В данной лабораторной работе необходимо загрузить и обработать датасет, а также выполнить задания по варианту с помощью библиотеки `pandas` в `Python`. Для начала была импортирована данная библиотека с помощью функции `import`, затем был загружен датасет `movies.csv` c помощью метода `read_csv`

In [248]:
import pandas as pd 

df=pd.read_csv('/Users/denis/Downloads/movies.csv', sep=';')

### 2. Обзор данных

2.1 Для первичного анализа датафрейма и проверки корректного импорта был использован метод `head()` c аргументом 20 для отображения первых 20 записей, благодаря чему была замечена ошибка импорта и добавлен параметр разделителя `sep` по точке с запятой в импорт

In [249]:
df.head(20)

title  release_year  score  \
0      David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                     Inception        2010.0    8.8   
2                                  Forrest Gump        1994.0    8.8   
3                                    Anbe Sivam        2003.0    8.7   
4                            Bo Burnham: Inside        2021.0    8.7   
5                           Saving Private Ryan        1998.0    8.6   
6                              Django Unchained        2012.0    8.4   
7                                        Dangal        2016.0    8.4   
8                        Bo Burnham: Make Happy        2016.0    8.4   
9                         Louis C.K.: Hilarious        2010.0    8.4   
10              Dave Chappelle: Sticks & Stones        2019.0    8.4   
11                                     3 Idiots        2009.0    8.4   
12                                 Black Friday        2004.0    8.4   
13                                 Super Deluxe        2019.0    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom        2015.0    8.3   
15                  Once Upon a Time in America        1984.0    8.3   
16                                  Taxi Driver        1976.0    8.3   
17                          Like Stars on Earth        2007.0    8.3   
18                            Bo Burnham: What.        2013.0    8.3   
19                            Full Metal Jacket        1987.0    8.3   

    number_of_votes  duration   main_genre main_production  
0           31180.0        83  documentary              GB  
1         2268288.0       148        scifi              GB  
2         1994599.0       142        drama              US  
3           20595.0       160       comedy              IN  
4           44074.0        87       comedy              US  
5         1346020.0       169        drama              US  
6         1472668.0       165      western              US  
7          180247.0       161       action              IN  
8           14356.0        60       comedy              US  
9           11973.0        84       comedy              US  
10          25687.0        65       comedy              US  
11         385782.0       170       comedy              IN  
12          20611.0       143        crime              IN  
13          13680.0       176     thriller              IN  
14          17710.0        98  documentary              UA  
15         342335.0       229        drama              US  
16         795222.0       113        crime              US  
17         188234.0       165        drama              IN  
18          11488.0        60       comedy              US  
19         723306.0       116        drama              GB

Затем было необходимо проанализировать типы данных, так как даже при визуальном аналазе видны ошибки (например, год имеет тип `float`), для этого был использован метод `info()`

In [250]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


2.3 Метод `.describe()` используется для получения статистического обзора данных, то есть подсчета, среднего, стандартного отклонения, максимального и минимального значение, квартериальных значений. Столбец `release_year` исключается из метода, так как там находится значение года


In [251]:
description = df.drop(columns='release_year').describe()
display(description)

score  number_of_votes    duration
count  386.000000     3.880000e+02  389.000000
mean     7.512176     1.391523e+05  123.352185
std      0.443870     2.365279e+05   28.304800
min      6.900000     1.013900e+04   28.000000
25%      7.100000     2.055375e+04  104.000000
50%      7.400000     4.606950e+04  122.000000
75%      7.800000     1.551585e+05  139.000000
max      9.000000     2.268288e+06  229.000000

### Выводы:

Среднее по рейтингу 7.51, что достаточно высокий средний балл, то есть фильмы в данном датасете достаточно качественные, стандартное отклонение 0.44, что мало, значит, разброс по оценкам небольшой, минимальная оценка - 6.9, то есть даже самая минимальная из них достаточно высокая, оценки по квантилям очень высокие, опять же доказывает качество фильмов (как минимум, по мнению оценщиков). Можно предположить, что в датасете популярные и общепризнанные работы

Среднее по количеству голосов 139152 - это большое число, что указывает на популярность контента, стандартное отклонение 236528 - сильный разброс, то есть могут быть фильмы как популярные, так и нет, аналогично и доказывает большая разница между средним и медианой. Максимум 2268288 — явный выброс, вероятно, блокбастер.

Среднее по продолжительности - 124 минуты, что является обычной продолжительностью фильмов, стандартное отклонение - 28 минут, что, в целом, нормально для полнометражных фильмов, минимальная продолжительность фильмов - 28, то есть в датасете также присутствуют короткометражки, оценки по квантилям также показывают, что примерно 25% фильмов в датасете относятся к короткометражным. Максимальная длительность фильма - 229 минуты, скорее всего, это была режисерская (расширенная) версия


 2.4 Оценка названий столбцов. 
 Предметной областью является журнал учёта фильмов, который хранит  такую информацию, как:
+ Название фильма
+ Год выхода
+ Оценка
+ Количество голосов 
+ Длительность 
+ Главный жанр
+ Страна производства 

Все эти характеристики отображены в столбцах данного датасета. Чтобы показать все столбцы, используется функция `columns`

In [252]:
df.columns

Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')

Названия столбцов было принято считать удовлетворительным, так как они несут в себе информацию о содержащихся данных, не имеют ошибок и неточностей

### 3. Проверка пропусков

3.1 Всего записей в этом датасете 389, но в столбцах title, release_year, score, number_of_voice наблюдаются меньше записей, так как там есть значения `NaN`. Чтобы вывести все записи, имеющие хотя бы одно нулевое значение, используется функция `.isnull()`. Функция `.any(axis=1)` указывает на поиск значений именно по строкам.

In [253]:
df_nan = df[df.isnull().any(axis=1)]
display(df_nan)



title  release_year  score  number_of_votes  duration main_genre  \
121    NaN        2017.0    7.8          25593.0        94     comedy   
141    NaN        2015.0    7.4          13703.0       101    musical   
174    NaN           NaN    NaN              NaN       121      drama   
187    NaN        2013.0    7.5          93314.0       128      drama   
214  Crazy           NaN    NaN         507878.0       118    romance   
226    NaN        1999.0    NaN         180532.0       127      drama   

    main_production  
121              US  
141              US  
174             NaN  
187              US  
214              US  
226              US

3.2 Было принято решение удалить все записи, содержащие в себе хотя бы одно значение `NaN`, так как в дальнейшем отсутствие помешает рассчетам и может привести к ошибкам при смене типов данных. За удаление строк с нулевыми значениями отвечает функция `.dropna()`. После этого была проведена проверка нулевых значений в столбцах с помощью метода `.isna()`,а метод `.sum()` показывает сумму таких значений. В результате нулевые значения отсутствуют

In [254]:
display(len(df_nan))
df = df.dropna()
df.isna().sum()
display(len(df))

6

383

3.3 Соответственно было исключено 6 записей из 389, что составляет 2% от изначального датасета

### 4. Проверка дубликатов

#### Проверка явных дубликатов

4.1 Нужно проверить датафрейм на наличие дубликатов, для этого используется метод `duplictated()`, также чтобы узнать количество дубликатов используется метод `sum()`

In [255]:
print(df.duplicated().sum())

2


4.2 Всего получилось 2 дубликата, которые необходимо удалить с помощью метода `drop_dublicates()`, а метод `reset_index` c параметром `drop=False` избавят от пропусков и создаст столбец `index`

In [256]:
df = df.drop_duplicates().reset_index(drop=False)
df.head()

index                                     title  release_year  score  \
0      0  David Attenborough: A Life on Our Planet        2020.0    9.0   
1      1                                 Inception        2010.0    8.8   
2      2                              Forrest Gump        1994.0    8.8   
3      3                                Anbe Sivam        2003.0    8.7   
4      4                        Bo Burnham: Inside        2021.0    8.7   

   number_of_votes  duration   main_genre main_production  
0          31180.0        83  documentary              GB  
1        2268288.0       148        scifi              GB  
2        1994599.0       142        drama              US  
3          20595.0       160       comedy              IN  
4          44074.0        87       comedy              US

4.3 Проверим удаление дубликатов

In [257]:
print(df.duplicated().sum())

0


4.4 Для проверки значений числовых столбцов использовано сравнение `<=0` и метод `sum()`, по итогу можно заметить, что неположительных значений нет

In [258]:
not_positive_count_score=(df['score']<=0).sum()
not_positive_count_release_year=(df['release_year']<=0).sum()
not_positive_count_number_of_votes=(df['number_of_votes']<=0).sum()
not_positive_count_duration=(df['duration']<=0).sum()

print(not_positive_count_score,not_positive_count_release_year,not_positive_count_number_of_votes,not_positive_count_duration)

0 0 0 0


#### Проверка неявных дубликатов

4.5 Затем были проверены столбцы `main_genre` и `main_production` на наличие записей с одинаковыми значениями, но разные по написанию (регистр или синоним) с помощью функции `.unique()`

In [259]:
print(df['main_genre'].unique())

df['main_production'].unique()

['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'COMEDY' 'dramaa' 'musical']


array(['GB', 'US', 'IN', 'UA', 'CD', 'TR', 'ES', 'AU', 'JP', 'ZA', 'HK',
       'DE', 'KR', 'CA', 'BE', 'NO', 'NZ', 'MX', 'FR', 'MW', 'TH', 'AR',
       'PS', 'HU', 'IT', 'CN', 'PL', 'KH', 'IE', 'BR', 'XX', 'LT', 'NL',
       'DK', 'ID'], dtype=object)

Как можно заметить, в столбце `main_production` повторов нет, а в столбце `main_genre` есть повторы:

+ `comedy` и `COMEDY` (разный регистр, но суть одна)
+ `drama` и `dramaa` (ошибка во втором названии)
+ `thriller ` и `horror` (смысл похожий, стоит оставить только 1 категорию)

Для изменений будет использован метод `replace()`

In [260]:
df['main_genre']=df['main_genre'].replace('COMEDY','comedy').replace('dramaa','drama').replace('thriller','horror')
print(df['main_genre'].unique())

['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'horror' 'war' 'fantasy' 'romance' 'animation' 'sports' 'musical']


4.7 Аналогично стоит проверить столбец `title`

In [261]:
df['title'].unique()

array(['David Attenborough: A Life on Our Planet', 'Inception',
       'Forrest Gump', 'Anbe Sivam', 'Bo Burnham: Inside',
       'Saving Private Ryan', 'Django Unchained', 'Dangal',
       'Bo Burnham: Make Happy', 'Louis C.K.: Hilarious',
       'Dave Chappelle: Sticks & Stones', '3 Idiots', 'Black Friday',
       'Super Deluxe', "Winter on Fire: Ukraine's Fight for Freedom",
       'Once Upon a Time in America', 'Taxi Driver',
       'Like Stars on Earth', 'Bo Burnham: What.', 'Full Metal Jacket',
       'Warrior', 'Drishyam', 'Queen', 'Paan Singh Tomar',
       'Cowspiracy: The Sustainability Secret', 'Virunga', 'PK',
       'BДЃhubali 2: The Conclusion', 'Monty Python and the Holy Grail',
       'Article 15', 'Miracle in Cell No. 7', '13th', 'Andhadhun',
       'Bill Burr: Paper Tiger', 'Udaan', 'How to Train Your Dragon',
       'Klaus', 'Swades', 'Minnal Murali', 'Rang De Basanti',
       'Seaspiracy', 'Rush', 'Hannah Gadsby: Nanette', 'Barfi!', 'Haider',
       'Zindagi Na Mile

4.8 По результатам проверки, повторов в этом столбце не обнаружено

### 5. Провека типов данных

5.1 Теперь необходимо изменить типы данных некоторых столбцов. Было решено поменять тип данных столбца `number_of_votes`

Сейчас он имеет тип данных `float`, но корректным типом будет `int`, так как количество голосов не может быть дробными

Для этого будет использован метод `astype()`

Также было решено изменить тип данных столбца `release_year`, сейчас он имеет тип `float`, а теперь он будет иметь тип `int`, что тоже не совсем корректно, но так как отсутствуют значения месяца и дня, то этого будет достаточно


In [262]:
df['number_of_votes']=df['number_of_votes'].astype(int)
df['release_year']=df['release_year'].astype(int)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 381 entries, 0 to 380
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   index            381 non-null    int64  
 1   title            381 non-null    object 
 2   release_year     381 non-null    int64  
 3   score            381 non-null    float64
 4   number_of_votes  381 non-null    int64  
 5   duration         381 non-null    int64  
 6   main_genre       381 non-null    object 
 7   main_production  381 non-null    object 
dtypes: float64(1), int64(4), object(3)
memory usage: 23.9+ KB


### 6. Группировка данных

#### Задание 1

*`Группировка - жанр и количество фильмов каждого main_production`*

6.1 Группировка происходит с помощью метода `groupby()`, необходимо сгруппировать данные по столбцам `main_genre` и `main_production`, вывести количество записей, соответствующее стране производства и жанру, выведя только столбец `index` c методом `count()`

In [263]:
df_group=df.groupby(['main_genre','main_production'])
group=df_group['index'].count()
display(group)

main_genre  main_production
action      DE                 1
            IN                 2
            US                 2
animation   JP                 1
            US                 2
                              ..
scifi       US                 2
sports      DE                 1
war         CD                 1
            US                 2
western     US                 7
Name: index, Length: 88, dtype: int64

Получилась таблица, которая хранит в себе количество фильмов для каждого жанра и страны производства

#### Задание 2

*`Группировка - жанр и количество фильмов каждого main_production.`
`Создать датафрейм. Переименовать столбец с количеством в “сount”.`
`Отсортировать по убыванию столбца “count”`*

6.2 Так как группировка точно такая же, то можно работать с таблицей из прошлого задания. Но теперь необходимо из этой таблицы создать датафрейм с помощью метода `pd.DataFrame()`. Самой таблице была присвоена переменная `group`, на основе которой был создан датафрейм. После этого столбец `index` был переименован в столбец `Count`.  Далее значения необходимо отсортировать с помощью метода `sort_values()`, для сортировки по убыванию пишется аргумент `ascending=False`

In [264]:
group_df=pd.DataFrame(group)
group_df=group_df.rename(columns={'index':'Count'})
group_df.sort_values(by='Count', ascending=False)

Count
main_genre main_production       
drama      US                  53
           IN                  48
horror     US                  23
           IN                  22
comedy     US                  21
...                           ...
drama      NZ                   1
fantasy    CN                   1
           ES                   1
           FR                   1
drama      LT                   1

[88 rows x 1 columns]

В результате получился датафрейм, отсортированный по количеству фильмов из определенной страны и жанра

#### Задание 3

*`Сводная таблица (pivot_table) - Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по
main_production. Отсортировать по убыванию рейтинга. Округлить до одного знака`*

6.3 Метод `.pivot_table()` используется для создания сводных таблиц, которые позволяют агрегировать данные и представлять их в удобном формате. Основные атрибуты метода:
+ `values`: Столбец или столбцы для агрегирования
+ `index`: Столбец или столбцы, которые будут использоваться в качестве индексов строк
+ `columns`: Столбец или столбцы, которые будут использоваться в качестве заголовков столбцов
+ `aggfunc`: Функция или список функций для агрегации

В данном задании необходимо создать сводную таблицу среднего рейтинга по main_production, поэтому из списка используются только атрибуты `values`, `index` и `aggfunc`. За значения берётся рейтинг, за индексы - страна производства и функция усреднения.
Также значения были отсортированы по убыванию и округлены до 1 знака с помощью метода `.round(1)`

In [265]:
pivot=df.pivot_table(values='score',index='main_production',aggfunc='mean')

pivot.sort_values(by='score',ascending=False).round(1)

score
main_production       
CD                 8.2
UA                 8.1
ZA                 8.1
TR                 7.8
NZ                 7.8
MX                 7.7
TH                 7.6
MW                 7.6
IN                 7.6
GB                 7.6
ES                 7.6
AU                 7.6
PS                 7.5
JP                 7.5
AR                 7.5
HK                 7.5
US                 7.5
KR                 7.4
BE                 7.4
DE                 7.4
NO                 7.4
CA                 7.4
HU                 7.2
CN                 7.2
FR                 7.2
PL                 7.2
BR                 7.2
IT                 7.2
KH                 7.2
LT                 7.1
XX                 7.1
IE                 7.1
NL                 7.0
ID                 6.9
DK                 6.9

Получена таблица, показывающая средний рейтинг фильмов по каждой стране

#### Задание 4

*`Сводная таблица (pivot_table) - медианный рейтинг (score) фильмов по
жанрам - столбцы и main_production- строки. Отсортировать по возрастанию
main_production.`*

6.4 В данном задании необходимо повторить прошлую сводную таблицу, но добавить еще столбцы main_genre, сортировка проводится по `main_production` по возрастанию

In [266]:
pivot_score=df.pivot_table(values='score',columns='main_genre',index='main_production',aggfunc='median')

display(pivot_score.sort_values(by='main_production',ascending=True).round(1))

main_genre       action  animation  comedy  crime  documentary  drama  \
main_production                                                         
AR                  NaN        NaN     7.5    NaN          NaN    NaN   
AU                  NaN        NaN     8.1    NaN          NaN    7.3   
BE                  NaN        NaN     NaN    NaN          7.9    NaN   
BR                  NaN        NaN     NaN    NaN          7.2    NaN   
CA                  NaN        NaN     7.1    NaN          NaN    7.2   
CD                  NaN        NaN     NaN    NaN          NaN    NaN   
CN                  NaN        NaN     NaN    NaN          NaN    7.2   
DE                  7.1        NaN     NaN    NaN          NaN    7.2   
DK                  NaN        NaN     NaN    6.9          NaN    NaN   
ES                  NaN        NaN     8.1    NaN          NaN    7.4   
FR                  NaN        NaN     7.2    NaN          NaN    7.1   
GB                  NaN        NaN     8.0    7.4          9.0    7.2   
HK                  NaN        NaN     NaN    NaN          NaN    7.5   
HU                  NaN        NaN     NaN    NaN          NaN    7.0   
ID                  NaN        NaN     NaN    NaN          NaN    NaN   
IE                  NaN        NaN     NaN    NaN          NaN    7.0   
IN                  8.2        NaN     7.6    7.4          NaN    7.6   
IT                  NaN        NaN     NaN    NaN          NaN    7.0   
JP                  NaN        7.1     7.6    NaN          NaN    7.4   
KH                  NaN        NaN     NaN    NaN          NaN    7.2   
KR                  NaN        NaN     NaN    NaN          NaN    7.6   
LT                  NaN        NaN     NaN    NaN          NaN    7.1   
MW                  NaN        NaN     NaN    NaN          NaN    7.6   
MX                  NaN        NaN     NaN    NaN          NaN    7.7   
NL                  NaN        NaN     NaN    NaN          NaN    7.0   
NO                  NaN        NaN     6.9    7.8          NaN    NaN   
NZ                  NaN        NaN     7.8    NaN          NaN    7.8   
PL                  NaN        NaN     NaN    NaN          NaN    NaN   
PS                  NaN        NaN     NaN    NaN          NaN    NaN   
TH                  NaN        NaN     NaN    NaN          NaN    NaN   
TR                  NaN        NaN     7.8    NaN          NaN    7.8   
UA                  NaN        NaN     NaN    NaN          8.1    NaN   
US                  7.4        7.2     7.3    7.4          7.6    7.4   
XX                  NaN        NaN     7.1    NaN          NaN    NaN   
ZA                  NaN        NaN     NaN    NaN          8.1    NaN   

main_genre       fantasy  horror  musical  romance  scifi  sports  war  \
main_production                                                          
AR                   NaN     NaN      NaN      NaN    NaN     NaN  NaN   
AU                   NaN     7.3      NaN      NaN    NaN     NaN  NaN   
BE                   NaN     7.0      NaN      NaN    NaN     NaN  NaN   
BR                   NaN     NaN      NaN      NaN    NaN     NaN  NaN   
CA                   NaN     NaN      NaN      NaN    8.0     NaN  NaN   
CD                   NaN     NaN      NaN      NaN    NaN     NaN  8.2   
CN                   7.2     NaN      NaN      NaN    NaN     NaN  NaN   
DE                   NaN     7.5      NaN      NaN    NaN     7.4  NaN   
DK                   NaN     NaN      NaN      NaN    NaN     NaN  NaN   
ES                   7.5     7.4      NaN      NaN    NaN     NaN  NaN   
FR                   7.5     7.0      NaN      NaN    NaN     NaN  NaN   
GB                   NaN     7.4      NaN      7.3    8.0     NaN  NaN   
HK                   NaN     NaN      NaN      NaN    NaN     NaN  NaN   
HU                   7.5     NaN      NaN      NaN    NaN     NaN  NaN   
ID                   NaN     6.9      NaN      NaN    NaN     NaN  NaN   
IE                   NaN     7.2      Na

Данная таблица показывает средний значения рейтинга фильмов по разным странам и жанрам,  не все страны снимают все жанры фильмов, поэтому некоторые значения заменены на `NaN`

### Вывод

В данной лабораторной работе была проделана работа по обработке датасета журнала фильмов, были исключены ненужные записи изменены типы данных на корректные, а также анализированы данные с помощью группировки и создания сводной таблицы. Были освоены основные методы библиотеки `pandas` и варианты их использования:

1. Для вывода информации о датасете были использованы методы `.info()`, `.describe()`
2. Для вывода информации об определённых аспектах датасета были использованы методы `df.column`, `.isna()`, `.isnull()`, `.unique()`, `.duplicated()`
3. Для замены данных в датасете были использованы методы `.rename()` - для столбцов или значений внутри столбцов, `.astype()` - для изменения типа данных.

Также были использованы методы `.groupby()` для группировки значений, `sort_values()` для их сортировки, `.pivot_table()` для создания сводной таблицы и `pd.DataFrame()` для создания датафрейма из таблицы.

При выполнении индивидуального задания были получены некоторые результаты группировок и сводных таблиц: самый большой средний рейтинг имеют фильмы из Республики Конго, так как они снимают только фильмы в военном жанре, имеющие высокие рейтинг, США - лидер по производству фильмов в жанре драма и комедия


### Дополнительное задание

**` 1. Создать столбец “Категория года выхода (release_year)” (с помощью
категоризации). Выделить минимум 3 категории (свежий, старый, средний),
фильтрацию для уровня года выбрать самостоятельно, аргументировать
выбор. Создать сводную таблицу: средний и медианный год выхода по
жанру и категории года выхода.`**

Для того, чтобы выделить 3 категории года, необходимо проанализировать перцентили, используется метод `describe()`

In [267]:
df.describe()

index  release_year       score  number_of_votes    duration
count  381.000000    381.000000  381.000000     3.810000e+02  381.000000
mean   193.249344   2010.997375    7.509186     1.362626e+05  123.530184
std    112.626295     10.862659    0.445084     2.350996e+05   28.528329
min      0.000000   1954.000000    6.900000     1.013900e+04   28.000000
25%     95.000000   2008.000000    7.100000     2.043000e+04  104.000000
50%    194.000000   2014.000000    7.400000     4.491700e+04  122.000000
75%    291.000000   2018.000000    7.800000     1.534630e+05  139.000000
max    386.000000   2022.000000    9.000000     2.268288e+06  229.000000

Как можно заметить, 25% значений находятся в  2008 году или раньше, медиана находится на 2014 году, 75% не позже 2018, а оставшиеся 25% - позже 2018. Таким образом, можно выделить 3 категории:

+ Старый - ниже 2008
+ Средний - между 2009 и 2018
+ Свежий - после 2018

Для категоризации будет использована функция `categorize_year`

In [268]:
def categorize_year (df):
    if df['release_year']<=2008:
        return 'Старый'
    if (df['release_year'] >2008 )& (df['release_year'] <=2018):
        return 'Средний'
    if df['release_year']>2018:
        return 'Свежий'
    

Затем необходимо создать столбец release_year_category

In [269]:
df['release_year_category']=df['release_year']



Далее с помощью метода `apply` необходимо применить функцию ко всем строкам данного столбца, для чего еще используется аргумент `axis=1`

In [270]:
df['release_year_category']=df.apply(categorize_year, axis=1)


Затем создается сводная таблица `year_mean`, с содержанием среднего года по жанру и категории года c помощью метода `pivot_table`

In [271]:
year_mean=df.pivot_table(values='release_year', index='main_genre', columns='release_year_category',aggfunc='mean' )

display(year_mean)

release_year_category       Свежий      Средний       Старый
main_genre                                                  
action                 2021.000000  2014.000000  1998.000000
animation              2021.000000  2012.500000          NaN
comedy                 2020.000000  2014.230769  1997.105263
crime                  2020.500000  2013.000000  1999.600000
documentary            2019.666667  2016.000000          NaN
drama                  2020.000000  2014.100000  1998.600000
fantasy                2020.333333  2014.500000  2006.000000
horror                 2019.461538  2015.142857  1998.631579
musical                2020.000000  2015.000000          NaN
romance                2020.500000  2015.100000  1995.250000
scifi                  2021.000000  2015.000000  1999.333333
sports                 2021.000000          NaN          NaN
war                            NaN  2014.500000  1961.000000
western                        NaN  2014.200000  1973.000000

Аналогично создается сводная таблица `year_median`, содержащая медианный год по жанру и категории года

In [272]:
year_median=df.pivot_table(values='release_year',index='main_genre',columns='release_year_category',aggfunc='median')

display(year_median)

release_year_category  Свежий  Средний  Старый
main_genre                                    
action                 2021.0   2015.0  1998.0
animation              2021.0   2012.5     NaN
comedy                 2019.5   2014.5  2003.0
crime                  2020.5   2012.5  2004.0
documentary            2019.5   2016.5     NaN
drama                  2020.0   2014.0  2001.0
fantasy                2021.0   2015.0  2006.0
horror                 2019.0   2016.0  2002.0
musical                2020.0   2015.0     NaN
romance                2020.5   2016.0  2001.0
scifi                  2021.0   2017.0  1997.0
sports                 2021.0      NaN     NaN
war                       NaN   2014.5  1961.0
western                   NaN   2015.0  1973.0

### Вывод по заданию
Большинство жанров фильмов в датасете появилось еще в прошлом веке и актуально до сих пор, однако музыкальные, спортивные и документальные жанры появились не так давно. Также в выборке отсутствуют военные и вестерн фильмы современного происхождения, скорее всего, это связано с тем, что их производство сократилось.




**` 18. Создать столбец “Категория рейтинга” (score) (с помощью категоризации).
Выделить минимум 3 категории (высокий, низкий, средний), фильтрацию
для уровня рейтинга выбрать самостоятельно, аргументировать выбор.
Выполнить фильтрацию: выбрать только записи с высокий и средний
рейтинг + топ 5 жанров с наибольшей популярностью (количество записей) год выхода в определенном диапазоне (диапазон подобрать
самостоятельно, чтобы было достаточное количество записей).Создать
сводные таблицы: средний, минимальный, максимальный и медианный
рейтинг по категории рейтинга и жанру `**

Для того, чтобы выделить 3 категории рейтинга, необходимо проанализировать перцентили, используется метод `describe()`

In [273]:
df.describe()

index  release_year       score  number_of_votes    duration
count  381.000000    381.000000  381.000000     3.810000e+02  381.000000
mean   193.249344   2010.997375    7.509186     1.362626e+05  123.530184
std    112.626295     10.862659    0.445084     2.350996e+05   28.528329
min      0.000000   1954.000000    6.900000     1.013900e+04   28.000000
25%     95.000000   2008.000000    7.100000     2.043000e+04  104.000000
50%    194.000000   2014.000000    7.400000     4.491700e+04  122.000000
75%    291.000000   2018.000000    7.800000     1.534630e+05  139.000000
max    386.000000   2022.000000    9.000000     2.268288e+06  229.000000

Как можно заметить, 25% значений с рейтингом до 7.1 медиана находится на 7.4 году, 75% не более 7.8, а оставшиеся 25% - больше 7.8. Таким образом, можно выделить 3 категории:

**Стоит обратить внимание, что категории выделяются только на основе датасета, так как в дейтсвительности оценка 7.1 - не низкая**

+ Низкий - ниже 7.1
+ Средний - между 7.1 и 7.8
+ Свежий - больше 7.8

Для категоризации будет использована функция `ReturnScore`

In [274]:
def ReturnScore(df):
    if (df['Scored'])<=7.1:
        return 'Низкий'
    
    if (df['Scored']>=7.1) & (df['Scored']<=7.8):
        return 'средний'
    if df['Scored']>=7.8:
        return 'высокий'



Затем необходимо создать столбец Scored

In [275]:
df['Scored']=df['score']

Далее с помощью метода `apply` необходимо применить функцию ко всем строкам данного столбца, для чего еще используется аргумент `axis=1`

In [276]:
df['Scored']=df.apply(ReturnScore, axis=1)

Далее нужно посчитать ТОП-5 жанров фильмов с наибольшей популярностью, для этого использован метод `value_counts()`

In [277]:
df['main_genre'].value_counts()

main_genre
drama          148
horror          67
comedy          57
crime           21
documentary     20
romance         20
fantasy         19
scifi            7
western          7
action           5
war              3
animation        3
musical          3
sports           1
Name: count, dtype: int64

Как можно заметить, лидируют drama, horror, comedy, crime. Documentry и romance имееют одинаковое количество записей, поэтому оба жанра попадают в топ-5

Далее необходимо отфильтровать данные с помощью метода `query`. Выбираются записи с высоким и средним рейтингом, года произведения между 2000 и 2025 и входящие в топ-5 жанров, определнных выше

In [278]:
df = df.query('(Scored == "средний" or Scored == "высокий") and '
              'release_year > 2000 and release_year < 2025 and '
              'main_genre in ["drama", "horror", "comedy", "crime", "documentry"]')

Далее создаются сводные таблицы с помощью метода `pivot_table`: средняя, максимальная, минимальная, медианная по категории рейтинга и жанра

In [279]:
df_mean=df.pivot_table(values="score",index="Scored",columns="release_year", aggfunc="mean")

df_max=df.pivot_table(values="score",index="Scored",columns="release_year", aggfunc="max")

df_min=df.pivot_table(values="score",index="Scored",columns="release_year", aggfunc="min")

df_median=df.pivot_table(values="score",index="Scored",columns="release_year", aggfunc="median")


In [280]:
df_mean

release_year      2001  2002  2003  2004  2005      2006   2007  2008  2009  \
Scored                                                                        
высокий       8.066667   8.1  8.20  8.25   NaN  8.033333  8.075  8.00  8.15   
средний       7.500000   7.7  7.56  7.50   7.3       NaN  7.340  7.56  7.56   

release_year      2010  ...      2013      2014  2015      2016     2017  \
Scored                  ...                                                
высокий       8.250000  ...  8.100000  8.100000  8.10  8.400000  8.00000   
средний       7.457143  ...  7.557143  7.466667  7.36  7.418182  7.40625   

release_year   2018      2019      2020  2021  2022  
Scored                                               
высокий       8.025  8.185714       NaN  8.20  8.00  
средний       7.500  7.480000  7.483333  7.45  7.25  

[2 rows x 22 columns]

In [281]:
df_max

release_year  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  ...  \
Scored                                                                    ...   
высокий        8.1   8.1   8.7   8.4   NaN   8.1   8.3   8.0   8.4   8.4  ...   
средний        7.6   7.7   7.8   7.8   7.3   NaN   7.5   7.7   7.6   7.8  ...   

release_year  2013  2014  2015  2016  2017  2018  2019  2020  2021  2022  
Scored                                                                    
высокий        8.3   8.2   8.2   8.4   8.1   8.2   8.4   NaN   8.7   8.0  
средний        7.8   7.8   7.7   7.8   7.8   7.7   7.8   7.7   7.6   7.3  

[2 rows x 22 columns]

In [282]:
df_min

release_year  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  ...  \
Scored                                                                    ...   
высокий        8.0   8.1   7.9   8.1   NaN   8.0   7.9   8.0   7.9   8.1  ...   
средний        7.4   7.7   7.3   7.2   7.3   NaN   7.2   7.4   7.4   7.3  ...   

release_year  2013  2014  2015  2016  2017  2018  2019  2020  2021  2022  
Scored                                                                    
высокий        8.0   8.0   8.0   8.4   7.9   7.9   7.9   NaN   7.9   8.0  
средний        7.2   7.2   7.2   7.2   7.2   7.2   7.2   7.2   7.4   7.2  

[2 rows x 22 columns]

In [283]:
df_median

release_year  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  ...  \
Scored                                                                    ...   
высокий        8.1   8.1   8.0  8.25   NaN   8.0  8.05   8.0  8.15  8.25  ...   
средний        7.5   7.7   7.6  7.50   7.3   NaN  7.40   7.6  7.60  7.40  ...   

release_year  2013  2014  2015  2016  2017  2018  2019  2020  2021  2022  
Scored                                                                    
высокий       8.05   8.1   8.1   8.4   8.0   8.0   8.2   NaN   8.0  8.00  
средний       7.50   7.5   7.3   7.3   7.4   7.6   7.4  7.55   7.4  7.25  

[2 rows x 22 columns]

### Выводы по заданию

Были выделены жанры фильмов, входяшие в ТОП-5 по количеству записей: drama, horror, comedy, crime, documentry. Анализ сводных таблиц дает понять, что категоризация была применена не ко всем годам, так как некоторые записи в датасете не подходят под условие. Также трудно вывести закономерность (нет такого, чтобы более свежие фильмы оценивали лучше или хуже, хотя качество съемки и т.д. меняется) Стоит отметить, что в 2002 году значения для средней категории составляют 7.7, учитывая, что условие для данной категоризации было между 7.1 и 7.8 - это хороший результат, показывающий, что все фильмы этого года близки к высокому рейтингу (при нашей выборке и условиях). В целом, как и было замечено при категоризации, рейтинг всех фильмов в выборке достаточно большой, что говорит о их качестве (как минимум, по мнению оценщиков)